File: src/projects/z4/z4_algcon.m
(* Z4d_algcon.m 
   Wolfgang Tichy 8/2009
*)

(* compute trA and detg, enforce algebraic constraints *)


(* variables *)
variables = {g[a,b], A[a,b],  trAstore, detgstore}


(* compute in this order *)
tocompute = {

  (* determinant of metric *)
  detg == matrixdet[g],
  Cinstruction == "if (detg<=0.) { PROBLEM; /*errorexit(\" detg <= 0.\");*/ detg=1.;} ",
  detginv == 1/detg,
  
  (* enforce detg = 1 *)
  Cif == normalizedetg,
    epsilon == detg-1.,
    Cinstruction == "if (fabs(epsilon)<1e-12) aux = 1.-oot*epsilon; else ",
    aux == detg^(-oot),
    g[a,b] == aux g[a,b],
  Cif == end,

  (* trace of extrinsic curvature using new metric if rescaled *)
  detg == matrixdet[g],
  Cinstruction == "if (detg<=0.) { PROBLEM; /*errorexit(\" detg <= 0.\");*/ detg=1.;} ",
  detginv == 1/detg,
  trA == detginv matrixinvdet[g,a,b] A[a,b],

  (* enforce trA = 0 *)
  Cif == subtractA,
    aux == -trA oot,
    A[a,b] == A[a,b] + aux g[a,b],
  Cif == end,

  trA == detginv matrixinvdet[g,a,b] A[a,b],

  (* store result if storage is enabled *)
  Cinstruction == "if (store) { ",
    detgstore  == detg,
    trAstore   == trA,
  Cinstruction == "} "
}




(* symmetries *)
 g[a_,b_] :=  g[b,a] /; !OrderedQ[{a,b}]
 A[a_,b_] :=  A[b,a] /; !OrderedQ[{a,b}]
ginv[a_,b_] := ginv[b,a] /; !OrderedQ[{a,b}]
Ainv[a_,b_] := Ainv[b,a] /; !OrderedQ[{a,b}]




(************************************************************************)
(* information for C output *)

(* information about the function 
   the C file will be functionname.c 
*)
includeANDdefine[] := Module[{},
  pr["#include \"bam.h\"\n"];
  pr["#include \"z4.h\"\n"];
  pr["\n"];
  pr["#define Power(x,y) (pow((double) (x), (double) (y)))\n"];
  pr["#define Log(x)     log((double) (x)))\n"];
  pr["#define pow2(x)    ((x)*(x))\n"];
  pr["#define pow2inv(x) (1.0/((x)*(x)))\n\n"];
  pr["#define Cal(x,y,z) ((x)?(y):(z))\n\n"];
  pr["#define PROBLEM    printf(\"  %d pts away from boundary\\n\",boundaryaway(6)); \\\n"];
  pr["                   printf(\"  %e %e %e \\n  detg=%e,trA=%e \\n\", \\\n"];
  pr["                   Ptr(level,\"x\")[ijk],Ptr(level,\"y\")[ijk],Ptr(level,\"z\")[ijk], \\\n"];
  pr["                   detg, trA); stop++ \n"];
  
  pr["\n\n\n"];
];

functionname = "z4_algcon";

functionbegin     = "bampi_openmp_start\n";

functionarguments = "tVarList *u";

functionloop = "int stop=0; forallpoints_ijk_openmp(level) {";

endfunctionloop = "} endfor_ijk_openmp;/* if (stop) errorexit(\"\"); */ /* loop ccc */";

endfunction       = "bampi_openmp_stop\n";


(* custom variable declaration
   we have to translate between the tensor names and the C variables  
   we may need special variables, for example for parameters
*)
variabledeclarations[] := Module[{},

  prdecvl[{g[a,b], A[a,b], trAstore, detgstore}, "u"];

  pr["tL *level = u->level;\n"];
  pr["\n"];

  pr["const int subtractA      = Getv(\"z4_subtractA\", \"yes\");\n"];
  pr["const int normalizedetg  = Getv(\"z4_normalizedetg\", \"yes\");\n"];
  pr["const double oot = 1.0/3.0;\n"];
  pr["const int store          = Getv(\"z4_register_algcon\", \"store\");\n"];
];    




(************************************************************************)
(* now we are ready to go *)

<< "../../math/MathToC/TensorEquationsToC.m"

File: src/projects/z4/z4_algcon.c
/* z4_algcon.c */
/* Copyright (C) 1998 Bernd Bruegmann, 24.1.2024 */
/* Produced with Mathematica */

#include "bam.h"
#include "z4.h"

#define Power(x,y) (pow((double) (x), (double) (y)))
#define Log(x)     log((double) (x)))
#define pow2(x)    ((x)*(x))
#define pow2inv(x) (1.0/((x)*(x)))

#define Cal(x,y,z) ((x)?(y):(z))

#define PROBLEM    printf("  %d pts away from boundary\n",boundaryaway(6)); \
                   printf("  %e %e %e \n  detg=%e,trA=%e \n", \
                   Ptr(level,"x")[ijk],Ptr(level,"y")[ijk],Ptr(level,"z")[ijk], \
                   detg, trA); stop++ 



void z4_algcon(tVarList *u)
{

double *g11 = vldataptr(u, 0);
double *g12 = vldataptr(u, 1);
double *g13 = vldataptr(u, 2);
double *g22 = vldataptr(u, 3);
double *g23 = vldataptr(u, 4);
double *g33 = vldataptr(u, 5);
double *A11 = vldataptr(u, 6);
double *A12 = vldataptr(u, 7);
double *A13 = vldataptr(u, 8);
double *A22 = vldataptr(u, 9);
double *A23 = vldataptr(u, 10);
double *A33 = vldataptr(u, 11);
double *trAstore = vldataptr(u, 12);
double *detgstore = vldataptr(u, 13);
tL *level = u->level;

const int subtractA      = Getv("z4_subtractA", "yes");
const int normalizedetg  = Getv("z4_normalizedetg", "yes");
const double oot = 1.0/3.0;
const int store          = Getv("z4_register_algcon", "store");



bampi_openmp_start


double dx = level->dx;
double dy = level->dy;
double dz = level->dz;
double oo2dx = 1/(2*dx), oo2dy = 1/(2*dy), oo2dz = 1/(2*dz);
double oodx2 = 1/(dx*dx), oody2 = 1/(dy*dy), oodz2 = 1/(dz*dz);
double oo4dxdy = 1/(4*dx*dy);
double oo4dxdz = 1/(4*dx*dz);
double oo4dydz = 1/(4*dy*dz);

double aux = 0.;
double detg = 0.;
double detginv = 0.;
double epsilon = 0.;
double trA = 0.;



int stop=0; forallpoints_ijk_openmp(level) {


detg
=
2.*g12[ijk]*g13[ijk]*g23[ijk] + g11[ijk]*g22[ijk]*g33[ijk] - 
  g33[ijk]*pow2(g12[ijk]) - g22[ijk]*pow2(g13[ijk]) - g11[ijk]*pow2(g23[ijk])
;


if (detg<=0.) { PROBLEM; /*errorexit(" detg <= 0.");*/ detg=1.;}  

detginv
=
1/detg
;



/* conditional */
if (normalizedetg) {

epsilon
=
-1. + detg
;


if (fabs(epsilon)<1e-12) aux = 1.-oot*epsilon; else  

aux
=
Power(detg,-oot)
;

g11[ijk]
=
aux*g11[ijk]
;

g12[ijk]
=
aux*g12[ijk]
;

g13[ijk]
=
aux*g13[ijk]
;

g22[ijk]
=
aux*g22[ijk]
;

g23[ijk]
=
aux*g23[ijk]
;

g33[ijk]
=
aux*g33[ijk]
;

}
/* if (normalizedetg) */


detg
=
2.*g12[ijk]*g13[ijk]*g23[ijk] + g11[ijk]*g22[ijk]*g33[ijk] - 
  g33[ijk]*pow2(g12[ijk]) - g22[ijk]*pow2(g13[ijk]) - g11[ijk]*pow2(g23[ijk])
;


if (detg<=0.) { PROBLEM; /*errorexit(" detg <= 0.");*/ detg=1.;}  

detginv
=
1/detg
;

trA
=
detginv*(-2.*A23[ijk]*g11[ijk]*g23[ijk] + A11[ijk]*g22[ijk]*g33[ijk] + 
    g11[ijk]*(A33[ijk]*g22[ijk] + A22[ijk]*g33[ijk]) + 
    2.*(g13[ijk]*(A23[ijk]*g12[ijk] - A13[ijk]*g22[ijk] + 
          A12[ijk]*g23[ijk]) + 
       g12[ijk]*(A13[ijk]*g23[ijk] - A12[ijk]*g33[ijk])) - 
    A33[ijk]*pow2(g12[ijk]) - A22[ijk]*pow2(g13[ijk]) - 
    A11[ijk]*pow2(g23[ijk]))
;



/* conditional */
if (subtractA) {

aux
=
-(oot*trA)
;

A11[ijk]
=
A11[ijk] + aux*g11[ijk]
;

A12[ijk]
=
A12[ijk] + aux*g12[ijk]
;

A13[ijk]
=
A13[ijk] + aux*g13[ijk]
;

A22[ijk]
=
A22[ijk] + aux*g22[ijk]
;

A23[ijk]
=
A23[ijk] + aux*g23[ijk]
;

A33[ijk]
=
A33[ijk] + aux*g33[ijk]
;

}
/* if (subtractA) */


trA
=
detginv*(-2.*A23[ijk]*g11[ijk]*g23[ijk] + A11[ijk]*g22[ijk]*g33[ijk] + 
    g11[ijk]*(A33[ijk]*g22[ijk] + A22[ijk]*g33[ijk]) + 
    2.*(g13[ijk]*(A23[ijk]*g12[ijk] - A13[ijk]*g22[ijk] + 
          A12[ijk]*g23[ijk]) + 
       g12[ijk]*(A13[ijk]*g23[ijk] - A12[ijk]*g33[ijk])) - 
    A33[ijk]*pow2(g12[ijk]) - A22[ijk]*pow2(g13[ijk]) - 
    A11[ijk]*pow2(g23[ijk]))
;


if (store) {  

detgstore[ijk]
=
detg
;

trAstore[ijk]
=
trA
;


}  



} endfor_ijk_openmp;/* if (stop) errorexit(""); */ /* loop ccc */



bampi_openmp_stop


}  /* function */

/* z4_algcon.c */
/* nvars = 14, nauxs = 5, n* = 124,  n/ = 37,  n+ = 50, n = 211, O = 0 */
